sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala

```diff
@@ -372,6 +372,11 @@ object JsonInferSchema {
         case (DoubleType, _: DecimalType) | (_: DecimalType, DoubleType) =>
           DoubleType
 
+        // This branch is only used by `SchemaOfVariant.mergeSchema` because `JsonInferSchema` never
+        // produces `FloatType`.
+        case (FloatType, _: DecimalType) | (_: DecimalType, FloatType) =>
```

Comment:
Is there a standard for this? These two are not compatible, because float is approximate but decimal is not.

Anyway, I'm fine with this as we already did it for double type.

Reply:
I'm not aware of an existing standard, but I think we can follow the type resolution rules in the add operator:
![image](https://github.com/apache/spark/assets/122326661/124461ab-fc5c-4470-848a-6a531b62a544)

Based on this result, I think it may be better to change the result of `float x decimal` into double.

Changing decimal into float/double may indeed lose precision, but I think it is a reasonable approach in the shcema inference.